File: tests/test_kage_reviewer_workflow.py
#!/usr/bin/env python3
"""
tests/test_kage_reviewer_workflow.py — Verifies Kage as mandatory Reviewer before final delivery.
Verifies:
1. Workflow transitions through: route -> explore (genin) -> plan (kage) -> execute (anbu/jonin) -> document (tokubetsu) -> review (kage).
2. Kage verifies that all tasks are 100% completed.
3. Kage rejects approval if any task is incomplete, has validation errors, or fails security checks.
4. Kage approves valid execution, enabling advancement to synthesize (sannin) -> done.
"""

import sys
import os
import tempfile
import unittest
import json

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))
import server

class TestKageReviewerWorkflow(unittest.TestCase):
    def setUp(self):
        self.tmp_dir = tempfile.TemporaryDirectory()
        self.task_dir = self.tmp_dir.name

    def tearDown(self):
        self.tmp_dir.cleanup()

    def test_kage_review_dispatch_when_in_review_phase(self):
        # Setup workflow in review phase
        status = {
            "phase": "review",
            "tasks": [
                {"id": "task-1", "agent": "anbu", "task": "Build backend", "status": "completed", "result": "Done", "validation": ["pass"]}
            ],
            "executed": {
                "task-1": {"agent": "anbu", "task": "Build backend", "result": "Done", "validation": ["pass"]}
            }
        }
        server._save_workflow_status(self.task_dir, status)

        # 1. Run workflow loop: Should dispatch Kage as reviewer
        res_json = server.run_mcp_workflow(task_dir=self.task_dir)
        res = json.loads(res_json)
        self.assertEqual(res["status"], "ready")
        self.assertEqual(res["phase"], "review")
        self.assertEqual(res["agent"], "kage")
        self.assertTrue(os.path.exists(os.path.join(self.task_dir, "delegate.md")))

    def test_kage_review_rejection_on_validation_errors(self):
        # 1. Dispatch Kage
        status = {
            "phase": "review",
            "tasks": [
                {"id": "task-1", "agent": "anbu", "task": "Build backend", "status": "completed", "result": "Done", "validation": ["pass"]}
            ]
        }
        server._save_workflow_status(self.task_dir, status)
        server.run_mcp_workflow(task_dir=self.task_dir)

        # 2. Write review with validation failure and result.md
        review_artifact = {
            "approved": False,
            "verified_task_ids": ["task-1"],
            "security_reviewed": True,
            "rollback_reviewed": True,
            "validation": ["FAIL: CVE vulnerability in dependency"]
        }
        with open(os.path.join(self.task_dir, "kage_review.json"), "w") as f:
            json.dump(review_artifact, f)
        with open(os.path.join(self.task_dir, "result.md"), "w") as f:
            f.write("Review completed with failures.")

        # 3. Advance workflow: Must be blocked
        res_json = server.run_mcp_workflow(task_dir=self.task_dir)
        res = json.loads(res_json)
        self.assertEqual(res["status"], "blocked")
        self.assertEqual(res["phase"], "review")
        self.assertIn("Kage review", res["message"])

    def test_kage_review_rejection_on_low_confidence(self):
        # 1. Dispatch Kage
        status = {
            "phase": "review",
            "tasks": [
                {"id": "task-1", "agent": "anbu", "task": "Build backend", "status": "completed", "result": "Done", "validation": ["pass"]}
            ]
        }
        server._save_workflow_status(self.task_dir, status)
        server.run_mcp_workflow(task_dir=self.task_dir)

        # 2. Write review with low confidence (85% < 90%)
        review_artifact = {
            "approved": True,
            "confidence": 85,
            "verified_task_ids": ["task-1"],
            "security_reviewed": True,
            "rollback_reviewed": True,
            "validation": ["all tests passed"]
        }
        with open(os.path.join(self.task_dir, "kage_review.json"), "w") as f:
            json.dump(review_artifact, f)
        with open(os.path.join(self.task_dir, "result.md"), "w") as f:
            f.write("Review completed with 85% confidence.")

        # 3. Advance workflow: Must be blocked because confidence < 90%
        res_json = server.run_mcp_workflow(task_dir=self.task_dir)
        res = json.loads(res_json)
        self.assertEqual(res["status"], "blocked")
        self.assertEqual(res["phase"], "review")

    def test_kage_review_approval_advances_to_synthesize(self):
        # 1. Dispatch Kage
        status = {
            "phase": "review",
            "tasks": [
                {"id": "task-1", "agent": "anbu", "task": "Build backend", "status": "completed", "result": "Done", "validation": ["pass"]}
            ],
            "executed": {
                "task-1": {"agent": "anbu", "task": "Build backend", "result": "Done", "validation": ["pass"]}
            }
        }
        server._save_workflow_status(self.task_dir, status)
        server.run_mcp_workflow(task_dir=self.task_dir)

        # Create a transient debug script to verify cleanup
        debug_script = os.path.join(self.task_dir, "debug_check.py")
        with open(debug_script, "w") as f:
            f.write("# temp debug script")

        # 2. Write clean approved review with 95% confidence and result.md
        review_artifact = {
            "approved": True,
            "confidence": 95,
            "verified_task_ids": ["task-1"],
            "security_reviewed": True,
            "rollback_reviewed": True,
            "validation": ["all 10 tests passed successfully"]
        }
        with open(os.path.join(self.task_dir, "kage_review.json"), "w") as f:
            json.dump(review_artifact, f)
        with open(os.path.join(self.task_dir, "result.md"), "w") as f:
            f.write("Kage review verified all tasks and security requirements.")

        # 3. Advance workflow: Must transition to synthesize (sannin)
        res_json = server.run_mcp_workflow(task_dir=self.task_dir)
        res = json.loads(res_json)
        self.assertEqual(res["status"], "completed")
        self.assertEqual(res["phase"], "done")
        self.assertTrue(os.path.exists(os.path.join(self.task_dir, "final_report.md")))
        # Verify transient debug file was cleaned up
        self.assertFalse(os.path.exists(debug_script))

if __name__ == '__main__':
    unittest.main()

File: tests/test_kage_reviewer_workflow.js
#!/usr/bin/env node
// DB isolation: keep test writes out of the production ~/.konoha/konoha.db
require('./helpers/isolate_db');

'use strict';

const assert = require('assert');
const fs = require('fs');
const os = require('os');
const path = require('path');
const server = require('../src/server');

async function run() {
  console.log('Running test_kage_reviewer_workflow tests...');
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'kage_rev_'));

  try {
    // 1. Kage review dispatch when in review phase
    const status1 = {
      phase: 'review',
      tasks: [
        { id: 'task-1', agent: 'anbu', task: 'Build backend', status: 'completed', result: 'Done', validation: ['pass'] }
      ],
      executed: {
        'task-1': { agent: 'anbu', task: 'Build backend', result: 'Done', validation: ['pass'] }
      }
    };
    server.saveWorkflowStatus(tmpDir, status1);
    const res1 = JSON.parse(server.runMcpWorkflow(tmpDir));
    assert.strictEqual(res1.status, 'ready');
    assert.strictEqual(res1.phase, 'review');
    assert.strictEqual(res1.agent, 'kage');
    assert.ok(fs.existsSync(path.join(tmpDir, 'delegate.md')));
    console.log('✓ Kage review dispatch passed');

    // 2. Kage review rejection on validation errors
    const status2 = {
      phase: 'review',
      tasks: [
        { id: 'task-1', agent: 'anbu', task: 'Build backend', status: 'completed', result: 'Done', validation: ['pass'] }
      ]
    };
    server.saveWorkflowStatus(tmpDir, status2);
    server.runMcpWorkflow(tmpDir);

    const reviewArtifact2 = {
      approved: false,
      verified_task_ids: ['task-1'],
      security_reviewed: true,
      rollback_reviewed: true,
      validation: ['FAIL: CVE vulnerability in dependency']
    };
    fs.writeFileSync(path.join(tmpDir, 'kage_review.json'), JSON.stringify(reviewArtifact2), 'utf8');
    fs.writeFileSync(path.join(tmpDir, 'result.md'), 'Review completed with failures.', 'utf8');

    const res2 = JSON.parse(server.runMcpWorkflow(tmpDir));
    assert.strictEqual(res2.status, 'blocked');
    assert.strictEqual(res2.phase, 'review');
    assert.ok(res2.message.includes('Kage review'));
    console.log('✓ Kage review rejection on validation errors passed');

    // 3. Kage review rejection on low confidence
    const status3 = {
      phase: 'review',
      tasks: [
        { id: 'task-1', agent: 'anbu', task: 'Build backend', status: 'completed', result: 'Done', validation: ['pass'] }
      ]
    };
    server.saveWorkflowStatus(tmpDir, status3);
    server.runMcpWorkflow(tmpDir);

    const reviewArtifact3 = {
      approved: true,
      confidence: 85,
      verified_task_ids: ['task-1'],
      security_reviewed: true,
      rollback_reviewed: true,
      validation: ['all tests passed']
    };
    fs.writeFileSync(path.join(tmpDir, 'kage_review.json'), JSON.stringify(reviewArtifact3), 'utf8');
    fs.writeFileSync(path.join(tmpDir, 'result.md'), 'Review completed with 85% confidence.', 'utf8');

    const res3 = JSON.parse(server.runMcpWorkflow(tmpDir));
    assert.strictEqual(res3.status, 'blocked');
    assert.strictEqual(res3.phase, 'review');

    // 3b. Verify 97% confidence is still blocked (threshold is strictly >= 98%)
    server.runMcpWorkflow(tmpDir);
    reviewArtifact3.confidence = 97;
    delete reviewArtifact3.categories;
    fs.writeFileSync(path.join(tmpDir, 'kage_review.json'), JSON.stringify(reviewArtifact3), 'utf8');
    fs.writeFileSync(path.join(tmpDir, 'result.md'), 'Review completed with 97% confidence.', 'utf8');
    const res3b = JSON.parse(server.runMcpWorkflow(tmpDir));
    assert.strictEqual(res3b.status, 'blocked');
    assert.strictEqual(res3b.phase, 'review');

    // 3c. Verify category-level confidence < 98 is blocked
    server.runMcpWorkflow(tmpDir);
    reviewArtifact3.confidence = 99;
    reviewArtifact3.categories = { 'Security Review': 97 };
    fs.writeFileSync(path.join(tmpDir, 'kage_review.json'), JSON.stringify(reviewArtifact3), 'utf8');
    fs.writeFileSync(path.join(tmpDir, 'result.md'), 'Review completed with category failure.', 'utf8');
    const res3c = JSON.parse(server.runMcpWorkflow(tmpDir));
    assert.strictEqual(res3c.status, 'blocked');
    assert.strictEqual(res3c.phase, 'review');
    console.log('✓ Kage review rejection on low confidence passed');

    // 4. Kage review approval advances to synthesize
    const status4 = {
      phase: 'review',
      tasks: [
        { id: 'task-1', agent: 'anbu', task: 'Build backend', status: 'completed', result: 'Done', validation: ['pass'] }
      ],
      executed: {
        'task-1': { agent: 'anbu', task: 'Build backend', result: 'Done', validation: ['pass'] }
      }
    };
    server.saveWorkflowStatus(tmpDir, status4);
    server.runMcpWorkflow(tmpDir);

    const debugScript = path.join(tmpDir, 'debug_check.py');
    fs.writeFileSync(debugScript, '# temp debug script', 'utf8');

    const reviewArtifact4 = {
      approved: true,
      confidence: 98,
      verified_task_ids: ['task-1'],
      security_reviewed: true,
      rollback_reviewed: true,
      ai_slop_findings: 0,
      ai_slop_clean: true,
      validation: ['all 10 tests passed successfully']
    };
    fs.writeFileSync(path.join(tmpDir, 'kage_review.json'), JSON.stringify(reviewArtifact4), 'utf8');
    fs.writeFileSync(path.join(tmpDir, 'result.md'), 'Kage review verified all tasks and security requirements.', 'utf8');

    const res4 = JSON.parse(server.runMcpWorkflow(tmpDir));
    assert.strictEqual(res4.status, 'completed');
    assert.strictEqual(res4.phase, 'done');
    assert.ok(fs.existsSync(path.join(tmpDir, 'final_report.md')));
    const finalReportContent = fs.readFileSync(path.join(tmpDir, 'final_report.md'), 'utf8');
    assert.ok(finalReportContent.includes('Minimum Required: ≥ 98%'), 'Report must specify Minimum Required: ≥ 98%');
    assert.ok(finalReportContent.includes('Minimum 98% required to allow delivery'), 'Report must specify Minimum 98% required');
    assert.strictEqual(fs.existsSync(debugScript), false);
    console.log('✓ Kage review approval advances to synthesize passed');

    console.log('\nAll test_kage_reviewer_workflow passed cleanly!');
  } finally {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  }
}

run().catch(err => {
  console.error(err);
  process.exit(1);
});
